### PR TITLE
Allow seperate configuration each RDS instance

### PR DESCRIPTION
**Pull request by mkatircioglu**
Fixes #gh-495

### CODE REVIEW

**Comment by mkatircioglu** [comment]
I haven't touched documentation part. I will do it finally according to your comments @maciejwalkowiak @eddumelendez

**Comment by mkatircioglu** [comment]
I just focused what the problem needs without refactoring too much. But if your suggestion is this way,  i will try to refactor my PR. @maciejwalkowiak

**Comment by maciejwalkowiak** [comment]
I don't have it 100% thought through but I don't see a reason anymore to not have configuration properties class for RDS integration. It would make the code itself easier to reason about but also easier to use for users thanks to properties autocompletion. It's something to try out and research and if you're willing to dive deeper to find the best option it would be helpful for sure.

**Comment by mkatircioglu** [comment]
I have provided configuration properties for AmazonRdsDatabaseAutoConfiguration but I am not sure how I can bind AnnotationAttributes to configuration properties for AmazonRdsInstanceConfiguration. Do you have any suggestion? I would be very happy, if you can give some guidance me. @maciejwalkowiak @eddumelendez

**Comment by maciejwalkowiak** [comment]
`spring-cloud-aws-jdbc` module is not Spring Boot aware, so it is not going to use configuration properties classes. Since we are deprecating `@EnableX` annotations (#591), `AmazonRdsInstanceConfiguration` will get deprecated too. Ideally, you separate auto-configuration completely from the configuration in `spring-cloud-aws-jdbc` module. Let me know if it's unclear. You have already done great job, thanks for that!

**Comment by mkatircioglu** [comment]
I have deprecated EnableRdsInstance and its autoconfiguration but I'm stuck on separating auto-configuration from spring-cloud-aws-jdbc. I might need some hints to go further more if you have a time @maciejwalkowiak

**Comment by maciejwalkowiak** [comment]
I'm on holidays this week. I'm happy to help when I'm back

**Comment by mkatircioglu** [comment]
Hi @maciejwalkowiak , I would be very happy If you have a time to guide me about on above comment.

**Comment by maciejwalkowiak** [comment]
@mkatircioglu thank you for your patience. If my plans work out I'll dig into your PR today/tomorrow.

**Comment by mkatircioglu** [comment]
Hi @maciejwalkowiak, Have you had a chance to check the PR?

**Comment by maciejwalkowiak** [comment]
Hi @mkatircioglu it took me a while, I was searching if there is a better way to register dynamically multiple beans. As far as my research goes - there is currently no way. Apologies for not giving you guidance but I had to dig through it myself to figure out how to do the separation - I've just commit it in a fork https://github.com/maciejwalkowiak/spring-cloud-aws/commit/1a262e9609c01d42aaabc2433674cf1bc76d2aa3. Except for tiny polishing I basically copy & pasted registering data source from the `jdbc` module - it's fine that they are deprecated as it will be gone from `jdbc` module anyway.

@mkatircioglu Take a look please at my commit and see if you find anything you would do differently.
@eddumelendez could you review this PR + my commit too?

**Comment by mkatircioglu** [comment]
The changes for polishing seems perfect for me. @maciejwalkowiak 

Btw, let me remind that the documentation changes are still missing which I can cover quickly if PR changes and your updates are fine for @eddumelendez too.

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ * @author Mete Alpaslan Katırcıoğlu
+ * @see org.springframework.cloud.aws.autoconfigure.jdbc.AmazonRdsDatabaseAutoConfiguration
+ */
+@ConfigurationProperties(prefix = "cloud.aws.rds")
+public class AmazonRdsDatabaseProperties {
+
+	/**
+	 * List of RdsInstances.
+	 */
+	private List<RdsInstance> instances = Collections.emptyList();
```
Comment (eddumelendez) [inline spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:38]: I would suggest to use `Map<String, RdsInstance>` so we can keep property configuration `cloud.aws.rds.<instance-name>` and keep backward compatibility

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
```
Comment (eddumelendez) [inline spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:2]: update copyright headers with `2020`

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		public boolean isReadReplicaSupport() {
+			return readReplicaSupport;
+		}
+
+		public void setReadReplicaSupport(boolean readReplicaSupport) {
+			this.readReplicaSupport = readReplicaSupport;
+		}
+
+		@Override
+		public boolean equals(Object o) {
```
Comment (eddumelendez) [inline spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:121]: the instance-name would be enough to know if the object is equal in this case

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(this.dbInstanceIdentifier, this.username,
+					this.databaseName, this.password, this.readReplicaSupport);
+		}
+
+		@Override
+		public String toString() {
```
Comment (eddumelendez) [inline spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:144]: is not needed

**Comment by mkatircioglu** [comment]
> @mkatircioglu Take a look please at my commit and see if you find anything you would do differently.

@maciejwalkowiak I will merge your polishing changes into my branch to continue on @eddumelendez suggestions? Is it fine or do you offer something else?

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java (hunk 1 above)
Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:38]: Actually, while adding configuration properties for RDS, the main reason of PR is lack of current RDS configuration and reshaping it to be more flexible as detailed in #gh-495

I'm not sure if we keep this backward compatible as it is. @eddumelendez

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java (hunk 2 above)
Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java]: Done!

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java (hunk 3 above)
Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:129]: Done!

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java (hunk 4 above)
Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java]: Done!

File: spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java (hunk 1 above)
Comment (eddumelendez) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: I think in yaml, list looks much better than properties. My suggestion was due to we can see a the properties assigned to one instance as you can see below:

```
cloud.aws.rds.my-instance.username=

cloud.aws.rds.my-instance.password=
```

or

```
cloud:
  aws:
    rds:
      my-instance:
        username:
        password:
```

wdyt?

Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: Sorry @eddumelendez but I got a bit confused. I don't understand what is the differences between what you offer and what is written in https://cloud.spring.io/spring-cloud-aws/spring-cloud-aws.html#_configuring_data_source_in_spring_boot.

Instead, following configuration changes is requested in #495  

```
cloud:
  aws:
    rds:
      - instanceName: ${ENV_VARIABLE_FOR_INSTANCE_ONE}
        databaseName: database-name-one
        username: user-one
        password: secret-one
```

and the changes helps to define `instanceName` as a separate environment variable as below.

```
cloud:
  aws:
    rds:
	  instances:
	  	- instanceName: ${ENV_VARIABLE_FOR_INSTANCE_ONE}
	      databaseName: database-name-one
	      username: user-one
	      password: secret-one
```

Comment (eddumelendez) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: @mkatircioglu we should also consider @maciejwalkowiak's opinion here

Let me explain, using `Map<String, RdsInstance>` as suggested we can use the same semantic as before `cloud.aws.rds.<instance-name>`. What does it mean?

Having, for example the configuration properties below

```
cloud.aws.rds.my-instance-1.username=foo
cloud.aws.rds.my-instance-1.password=bar

cloud.aws.rds.my-instance-2.username=a
cloud.aws.rds.my-instance-2.password=b
```

Will create a map with 2 entries, one for `my-instance-1` and other for `my-instance-2` with the given `username` and `password`.

At first glance, looks like haven't change anything but the magic is done by the configurarion properties class that you are introducing

and it is only a suggestion :)

Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: I got your point @eddumelendez. 

Btw, are you also sure that dbInstanceIdentifier can be replaceable by an environment variable if we use `Map<String, RdsInstance>`in `AmazonRdsDatabaseProperties` as described in below?

```
cloud.aws.rds.${ENV_VARIABLE_FOR_INSTANCE_ONE}.username=foo
cloud.aws.rds.${ENV_VARIABLE_FOR_INSTANCE_ONE}.password=bar

cloud.aws.rds.${ENV_VARIABLE_FOR_INSTANCE_TWO}.username=a
cloud.aws.rds.${ENV_VARIABLE_FOR_INSTANCE_TWO}.password=b

ENV_VARIABLE_FOR_INSTANCE_ONE=my-instance-1
ENV_VARIABLE_FOR_INSTANCE_TWO=my-instance-2
```

If so, I can change as you suggested.

Btw, what is your opinion @maciejwalkowiak?

Comment (maciejwalkowiak) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: AFAIR `cloud.aws.rds.${ENV_VARIABLE_FOR_INSTANCE_ONE}.username=foo` will not work and making it possible to externalize the instance id was the main motivation for this change.

Comment (eddumelendez) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: TBH, I am not sure about it :/

Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: According to your final comments, what is your latest decision? @eddumelendez @maciejwalkowiak

Comment (eddumelendez) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: I will go with the current (yours) implementation 👍🏽

I didn't take into account the environment variable resolution

Comment (mkatircioglu) [reply spring-cloud-aws-autoconfigure/src/main/java/org/springframework/cloud/aws/autoconfigure/jdbc/AmazonRdsDatabaseProperties.java:41]: Great! So, there is no more review comments and I'm resolving this! Thanks a lot @eddumelendez

**Comment by mkatircioglu** [comment]
Can we close this PR if you don't have any further comments? @eddumelendez @maciejwalkowiak

**Comment by maciejwalkowiak** [comment]
I'll do it today 👍

**Pull request closed without merging**